From f74f422f9b43c0bb99696de0c4dbe6306295c8ec Mon Sep 17 00:00:00 2001
From: Nicolas Dufresne <nicolas.dufresne@collabora.com>
Date: Thu, 6 Jul 2017 17:20:56 -0400
Subject: [PATCH 43/80] kmssink: Don't leak GEM primed from DMABuf

This otherwise breaks DMABuf reclaiming. This is not visible from
userspace, but inside the kernel, the DRM driver will hold a ref to the
DMABuf object. With a V4L2 driver allocating those DMABuf, it then
prevent changing the resolution and re-allocation new buffers.

https://bugzilla.gnome.org/show_bug.cgi?id=782774
---
 sys/kms/gstkmsallocator.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/sys/kms/gstkmsallocator.c b/sys/kms/gstkmsallocator.c
index a3c444396..18a540d43 100644
--- a/sys/kms/gstkmsallocator.c
+++ b/sys/kms/gstkmsallocator.c
@@ -483,6 +483,18 @@ gst_kms_allocator_dmabuf_import (GstAllocator * allocator, gint * prime_fds,
   if (!gst_kms_allocator_add_fb (alloc, tmp, offsets, vinfo))
     goto failed;
 
+  for (i = 0; i < n_planes; i++) {
+    struct drm_gem_close arg = { tmp->gem_handle[i], };
+    gint err;
+
+    err = drmIoctl (alloc->priv->fd, DRM_IOCTL_GEM_CLOSE, &arg);
+    if (err)
+      GST_WARNING_OBJECT (allocator,
+          "Failed to close GEM handle: %s %d", strerror (errno), errno);
+
+    tmp->gem_handle[i] = 0;
+  }
+
   return tmp;
 
   /* ERRORS */
-- 
2.11.0

